kernel/work.c

```diff
@@ -0,0 +1,1030 @@
+	}
+
+	k_spin_unlock(&lock, key);
+}
+
+void k_work_init_delayable(struct k_work_delayable *dwork,
+			    k_work_handler_t handler)
+{
+	__ASSERT_NO_MSG(dwork != NULL);
+	__ASSERT_NO_MSG(handler != 0);
+
+	*dwork = (struct k_work_delayable){
+		.work = {
+			.handler = handler,
+			.flags = K_WORK_DELAYABLE,
+		},
+	};
+	z_init_timeout(&dwork->timeout);
+	(void)work_timeout;
+}
+
+static inline int work_delayable_busy_get_locked(const struct k_work_delayable *dwork)
+{
+	return atomic_get(&dwork->work.flags) & K_WORK_MASK;
+}
+
+int k_work_delayable_busy_get(const struct k_work_delayable *dwork)
+{
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	int ret = flags_get(&dwork->work.flags) & K_WORK_MASK;
+
+	k_spin_unlock(&lock, key);
+	return ret;
+}
+
+/* Attempt to schedule a work item for future (maybe immediate)
+ * submission.
+ *
+ * Invoked with work lock held.
+ *
+ * See also submit_to_queue_locked(), which implements this for a no-wait
+ * delay.
+ *
+ * Invoked with work lock held.
+ *
+ * @param queuep pointer to a pointer to a queue.  On input this
+ * should dereference to the proposed queue (which may be null); after
+ * completion it will be null if the work was not submitted or if
+ * submitted will reference the queue it was submitted to.  That may
+ * or may not be the queue provided on input.
+ *
+ * @param dwork the delayed work structure
+ *
+ * @param delay the delay to use before scheduling.
+ *
+ * @retval from submit_to_queue_locked() if delay is K_NO_WAIT; otherwise
+ * @retval 1 to indicate successfully scheduled.
+ */
+static int schedule_for_queue_locked(struct k_work_q **queuep,
+				     struct k_work_delayable *dwork,
+				     k_timeout_t delay)
+{
+	int ret = 1;
+	struct k_work *work = &dwork->work;
+
+	if (K_TIMEOUT_EQ(delay, K_NO_WAIT)) {
+		return submit_to_queue_locked(work, queuep);
+	}
+
+	flag_set(&work->flags, K_WORK_DELAYED_BIT);
+	dwork->queue = *queuep;
+
+	/* Add timeout */
+	z_add_timeout(&dwork->timeout, work_timeout, delay);
+
+	return ret;
+}
+
+/* Unschedule delayable work.
+ *
+ * If the work is delayed, cancel the timeout and clear the delayed
+ * flag.
+ *
+ * Invoked with work lock held.
+ *
+ * @param dwork pointer to delayable work structure.
+ *
+ * @return true iff work had been delayed
+ */
+static inline bool unschedule_locked(struct k_work_delayable *dwork)
+{
+	bool ret = false;
+	struct k_work *work = &dwork->work;
+
+	/* If scheduled, try to cancel. */
+	if (flag_test_and_clear(&work->flags, K_WORK_DELAYED_BIT)) {
+		z_abort_timeout(&dwork->timeout);
+		ret = true;
+	}
+
+	return ret;
+}
+
+/* Full cancellation of a delayable work item.
+ *
+ * Unschedules the delayed part then delegates to standard work
+ * cancellation.
+ *
+ * Invoked with work lock held.
+ *
+ * @param dwork delayable work item
+ *
+ * @return k_work_busy_get() flags
+ */
+static int cancel_delayable_async_locked(struct k_work_delayable *dwork)
+{
+	(void)unschedule_locked(dwork);
```

Comment:
couldn't return value be used to avoid calling `cancel_async_locked` when timeout was cancelled?

Reply:
No, because the item might have been both scheduled and submitted.  `unschedule_locked` can only remove the scheduled submission; if it's already submitted or running it still needs to be cancelled normally.